### PR TITLE
Support asynchronous `SourceInputStreamFactory`

### PR DESCRIPTION
**Pull request by arteam**
In order to support async `RangeMissingHandler` callbacks in https://github.com/elastic/elasticsearch/pull/110587, we need to make sure that `SourceInputStreamFactory` provides `InputStream`s asynchronously first

### CODE REVIEW

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 1]
```diff
@@ -1339,9 +1339,9 @@ public interface SourceInputStreamFactory extends Releasable {
         /**
          * Create the input stream at the specified position.
          * @param relativePos the relative position in the remote storage to read from.
-         * @return the input stream ready to be read from.
+         * @param listener listener for the input stream ready to be read from
          */
-        InputStream create(int relativePos) throws IOException;
+        void create(int relativePos, ActionListener<InputStream> listener) throws IOException;
```
Comment (ywangd) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:1344]: Its [usage](https://github.com/elastic/elasticsearch/blob/38bcf07379393e048b3093a04df868e6ed0ff6e9/x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java#L994-L998) in `SharedBlobCacheService` assumes its synchronous, i.e. the try-with-resource will close the factory (in turn the underlying inputStream) when the `gapFillingTasks` finishes. But if the `fillCacheRange` (called by `gapFillingTasks`) is async, the factory will be closed too early? I think we need to remove the factory closing from the caller here and let the implementation of the factory and `fillCacheRange` handle it. What do you think?

**Review by ywangd — COMMENTED** [review]
Left a question.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java (hunk 1 above)
Comment (arteam) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:1345]: That's a really good point! I've made some changes and wonder if can piggy back on the `completionListener` on `fillGapRunnable` and close the `streamFactory` when all `fillCacheRange` methods have completed.

Comment (ywangd) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:1345]: Cool that looks better than I suggested.

**Review by kingherc — APPROVED** [review]
LGTM. Please merge after making sure the related serverless PR is also reviewed and OK to merge.

**Comment by arteam** [comment]
This change has been incorporated to #110587

**Pull request closed without merging**